### PR TITLE
seafile-seahub,django: cleanup seafile a bit, add python3 variants for django & django-constance

### PR DESCRIPTION
**Pull request by commodo**
Maintainer: me
Compile tested: http://github.com/openwrt/openwrt/commit/3209f5ae3d54e6f028e2e54314180a1be5b4896e x96
Run tested: http://github.com/openwrt/openwrt/commit/3209f5ae3d54e6f028e2e54314180a1be5b4896e x96

----------------------------------------------------------------------------

Description:

This change mostly focuses on decoupling django-admin from seafile-seahub.
And re-organizing some dependencies in a cleaner way.

A side effect is to also add python3-django. and python3-django-constance.
This is just a start, since all django packages will be converted.
The idea is to have the switch to Python3 (for seafile-seahub) be a one-liner if possible.

Signed-off-by: Alexandru Ardelean <ardeleanalex@gmail.com>

### CODE REVIEW

**Comment by commodo** [comment]
@cotequeiroz 
i think you care a bit about seafile

**Comment by cotequeiroz** [comment]
I'll run-test seafile (it may take a while), and post an approval review if OK.

**Comment by commodo** [comment]
ack

File: net/seafile-seahub/Makefile [hunk 1]
```diff
@@ -22,17 +22,20 @@ PKG_BUILD_DEPENDS:=python-pytz/host
 include $(INCLUDE_DIR)/package.mk
 include ../../lang/python/python-package.mk
 
+SEAFILE_PYTHON_DEPENDS:= \
+	simplejson chardet dateutil mysql pytz qrcode requests requests-oauthlib
```
Comment (neheb) [inline net/seafile-seahub/Makefile]: chardet can be removed (implied by requests). pytz by django,

File: net/seafile-seahub/Makefile [hunk 2]
```diff
@@ -22,17 +22,20 @@ PKG_BUILD_DEPENDS:=python-pytz/host
...
+
 define Package/seafile-seahub
     SECTION:=net
     CATEGORY:=Network
     TITLE:=Seafile server - seahub component
     URL:=https://seafile.com/
-    DEPENDS:=+python-simplejson +python +pillow +python-chardet +django +django-appconf \
+    DEPENDS:=+python +pillow +django +django-appconf \
```
Comment (neheb) [inline net/seafile-seahub/Makefile]: Pillow is implied by python-qrcode.

File: lang/python/django/Makefile [hunk 3]
```diff
@@ -21,46 +21,51 @@ PKG_LICENSE:=BSD-3-Clause
 PKG_LICENSE_FILES:=LICENSE LICENSE.python
 PKG_CPE_ID:=cpe:/a:djangoproject:django
 
+PKG_BUILD_DIR:=$(BUILD_DIR)/$(BUILD_VARIANT)-django-$(PKG_VERSION)
+
 include $(INCLUDE_DIR)/package.mk
 include ../python-package.mk
+include ../python3-package.mk
```
Comment (neheb) [inline lang/python/django/Makefile]: Would also be a great time to update the package as the latest point release has a CVE fix.

**Comment by neheb** [comment]
Something else that needs looking into is when taking a clean config and selecting seafile, libpq (PostgreSQL), libsqlite3, and mariadb all get selected even though they do the same thing.

**Comment by jefferyto** [comment]
I think double-checking dependencies is a good idea, but I would suggest staying close to upstream's [declared requirements](https://github.com/haiwen/seahub/blob/master/requirements.txt), even if some of them are implied by other packages.

File: lang/python/django/Makefile (hunk 3 above)
Comment (commodo) [reply lang/python/django/Makefile]: I would do that in an upcoming series.
There's plenty of things to change here.
And plenty to test & re-test. And I am not quite doing a great job here (yet).

But, if doing things incrementally, I can do a decent job and get better.

**Comment by commodo** [comment]
> Something else that needs looking into is when taking a clean config and selecting seafile, libpq (PostgreSQL), libsqlite3, and mariadb all get selected even though they do the same thing.

ack
this will require more work;
i don't really understand the scope of changes for this;
all i know is that seafile was added in a not-very-{clean,sane} manner;
making it sane, will be done incrementally

**Comment by cotequeiroz** [comment]
> 
> 
> I think double-checking dependencies is a good idea, but I would suggest staying close to upstream's [declared requirements](https://github.com/haiwen/seahub/blob/master/requirements.txt), even if some of them are implied by other packages.

+1
I don't think there's much value in "cleaning up" dependencies if they are actually used in code.  A dependent package can always switch its own dependencies, and removing them here will only create the need for a recursive (I don't mean circular) dependency check when you update a library, for example.

Say we have packA that "imports" PackB PackC PackD.  PackB also imports PackC, so we decide to remove PackC from PackA dependencies, and we do that treewide.  Then, all of a sudden, PackB no longer requires PackC, but starts to require PackD.  Sure PackC's absence **may** be visible (for python packages, for example, missing depends may only show at run-time when you reach that part of the code); PackD's "wrong" presence will not.  Do we want to have to search for all packages that have both PackB and PackD, so we can remove PackD for consistency?  The same applies to PackC: to keep consistency, we have to check all of the other PackA dependencies to ensure none of them depend on PackC before we can add PackC to PackA's DEPENDS.

To me, the "right thing to do" is to list the package's declared dependencies.  Sure, upstream may be cutting them down, but that's their decision (they may not have a multitude of packages to check), and it will become a bug if the above situation happens.

**Comment by neheb** [comment]
As is, probably needs to be rebased against master to pass CircleCI. Not sure if this is even complete or not.

**Comment by commodo** [comment]
rebased;
this isn't complete, it's a partial work;
more will be done later;

File: lang/python/django/Makefile [hunk 4]
```diff
@@ -21,46 +21,51 @@ PKG_LICENSE:=BSD-3-Clause
...
-	    $(PKG_INSTALL_DIR)$(PYTHON_PKG_DIR)/* \
-	    $(1)$(PYTHON_PKG_DIR)
-	$(INSTALL_BIN) $(PKG_INSTALL_DIR)/usr/bin/django-admin.py \
-	    $(1)/usr/bin
+define Package/python3-django
+$(call Package/django/Default)
+  DEPENDS:= \
+	+PACKAGE_python3-django:python3 \
+	+PACKAGE_python3-django:python3-pytz
+  VARIANT:=python
```
Comment (cotequeiroz) [inline lang/python/django/Makefile]:
```suggestion
  VARIANT:=python3
```
This is probably causing the build failure.

File: lang/python/django/Makefile [hunk 5]
```diff
@@ -21,46 +21,51 @@ PKG_LICENSE:=BSD-3-Clause
...
-	$(INSTALL_DIR) $(1)$(PYTHON_PKG_DIR)
-	$(INSTALL_DIR) $(1)/usr/bin
-	$(CP) \
-	    $(PKG_INSTALL_DIR)$(PYTHON_PKG_DIR)/* \
-	    $(1)$(PYTHON_PKG_DIR)
-	$(INSTALL_BIN) $(PKG_INSTALL_DIR)/usr/bin/django-admin.py \
-	    $(1)/usr/bin
-	# fix python exec path
-	sed -i 's/#!.*/#!\/usr\/bin\/python/g' $(1)/usr/bin/django-admin.py
+define Package/pytho3-django/description
```
Comment (cotequeiroz) [inline lang/python/django/Makefile]:
```suggestion
define Package/python3-django/description
```

**Comment by commodo** [comment]
Changelog v1 -> v2:
* rebased

Changelog v2 -> v3:
* fixed typos
* updated PKG_RELEASE for all packages modified

**Comment by neheb** [comment]
I think this was mentioned elsewhere. Seafile has a Python3 version now. Anyway, let's merge for now.

**Pull request merged**